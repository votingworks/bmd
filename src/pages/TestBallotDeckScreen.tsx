import React, { PointerEventHandler, useState } from 'react'

import { CandidateContest, Election, VotesDict } from '../config/types'

import Button from '../components/Button'
import ButtonList from '../components/ButtonList'
import Main, { MainChild } from '../components/Main'
import MainNav from '../components/MainNav'
import Prose from '../components/Prose'
import PrintedBallot from '../components/PrintedBallot'

interface Ballot {
  ballotId?: string
  precinctId: string
  ballotStyleId: string
  votes: VotesDict
}

interface GenerateTestDeckParams {
  election: Election
  precinctId?: string
}

const generateTestDeckBallots = ({
  election,
  precinctId,
}: GenerateTestDeckParams) => {
  const precincts: string[] = precinctId
    ? [precinctId]
    : election.precincts.map(p => p.id)

  let ballots: Ballot[] = []

  precincts.forEach(precinctId => {
    const precinct = election.precincts.find(p => p.id === precinctId)!
    const precinctBallotStyles = election.ballotStyles.filter(bs =>
      bs.precincts.includes(precinct.id)
    )

    precinctBallotStyles.forEach(ballotStyle => {
      const contests = election.contests.filter(
        c =>
          ballotStyle.districts.includes(c.districtId) &&
          ballotStyle.partyId === c.partyId
      )

      const numBallots = Math.max(
        ...contests.map(c =>
          c.type === 'yesno' ? 2 : (c as CandidateContest).candidates.length
        )
      )

      for (let ballotNum = 0; ballotNum < numBallots; ballotNum++) {
        let votes: VotesDict = {}
        contests.forEach(contest => {
          if (contest.type === 'yesno') {
            votes[contest.id] = ballotNum % 2 === 0 ? 'yes' : 'no'
          } else {
            if (contest.candidates.length > 0) {
              votes[contest.id] = [
                contest.candidates[ballotNum % contest.candidates.length],
              ]
            }
          }
        })
        ballots.push({
          ballotStyleId: ballotStyle.id,
          precinctId,
          votes,
        })
      }
    })
  })

  return ballots
}

interface Precinct {
  name: string
  id: string
}

interface Props {
  election: Election
  hideTestDeck: () => void
  isLiveMode: boolean
}

const initialPrecinct: Precinct = { id: '', name: '' }

const TestBallotDeckScreen = ({
  election,
  hideTestDeck,
  isLiveMode,
}: Props) => {
  const [ballots, setBallots] = useState<Ballot[]>([])
  const [precinct, setPrecinct] = useState<Precinct>(initialPrecinct)

  const selectPrecinct: PointerEventHandler = event => {
    const { id = '', name = '' } = (event.target as HTMLElement).dataset
    setPrecinct({ name, id })
    const selectedBallots = generateTestDeckBallots({
      election,
      precinctId: id,
    })
    setBallots(selectedBallots)
  }

  const resetDeck = () => {
    setBallots([])
    setPrecinct(initialPrecinct)
  }

  return (
    <React.Fragment>
      <Main>
        <MainChild maxWidth={false}>
          {ballots.length ? (
            <React.Fragment>
              <Prose className="no-print">
                <h1>Test Ballots Deck for {precinct.name}</h1>
                <p>
                  <Button primary onPress={window.print}>
                    Print {ballots.length} ballots
                  </Button>
                </p>
                <p>
                  <Button small onPress={resetDeck}>
                    Back to All Decks
                  </Button>
                </p>
              </Prose>
              {ballots.map((ballot, i) => {
                const ballotId = `temp-ballot-id-${i}`
                return (
                  <PrintedBallot
                    key={ballotId}
                    ballotId={undefined} // TODO: add ballotId here: https://github.com/votingworks/bmd/issues/424
                    ballotStyleId={ballot.ballotStyleId}
                    election={election}
                    isLiveMode={isLiveMode}
                    precinctId={ballot.precinctId}
                    votes={ballot.votes}
                  />
                )
              })}
            </React.Fragment>
          ) : (
            <React.Fragment>
              <Prose>
                <h1>Test Ballot Decks</h1>
                <p>
                  Select desired precinct for <strong>{election.title}</strong>.
                </p>
              </Prose>
              <p>
                <Button
                  data-id=""
                  data-name="All Precincts"
                  fullWidth
                  key="all-precincts"
                  onPress={selectPrecinct}
                >
                  <strong>All Precincts</strong>
                </Button>
              </p>
              <ButtonList>
                {election.precincts.map(p => (
                  <Button
                    data-id={p.id}
                    data-name={p.name}
                    fullWidth
                    key={p.id}
                    onPress={selectPrecinct}
                  >
                    {p.name}
                  </Button>
                ))}
              </ButtonList>
            </React.Fragment>
          )}
        </MainChild>
      </Main>
      <MainNav title="Clerk Actions">
        <Button small onPress={hideTestDeck}>
          Dashboard
        </Button>
      </MainNav>
    </React.Fragment>
  )
}

export default TestBallotDeckScreen
