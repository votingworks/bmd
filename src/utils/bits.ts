import { TextEncoder, TextDecoder } from 'text-encoding'

export type Uint8Index = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7
export type Uint1 = 0 | 1
export type Uint8 =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | 32
  | 33
  | 34
  | 35
  | 36
  | 37
  | 38
  | 39
  | 40
  | 41
  | 42
  | 43
  | 44
  | 45
  | 46
  | 47
  | 48
  | 49
  | 50
  | 51
  | 52
  | 53
  | 54
  | 55
  | 56
  | 57
  | 58
  | 59
  | 60
  | 61
  | 62
  | 63
  | 64
  | 65
  | 66
  | 67
  | 68
  | 69
  | 70
  | 71
  | 72
  | 73
  | 74
  | 75
  | 76
  | 77
  | 78
  | 79
  | 80
  | 81
  | 82
  | 83
  | 84
  | 85
  | 86
  | 87
  | 88
  | 89
  | 90
  | 91
  | 92
  | 93
  | 94
  | 95
  | 96
  | 97
  | 98
  | 99
  | 100
  | 101
  | 102
  | 103
  | 104
  | 105
  | 106
  | 107
  | 108
  | 109
  | 110
  | 111
  | 112
  | 113
  | 114
  | 115
  | 116
  | 117
  | 118
  | 119
  | 120
  | 121
  | 122
  | 123
  | 124
  | 125
  | 126
  | 127
  | 128
  | 129
  | 130
  | 131
  | 132
  | 133
  | 134
  | 135
  | 136
  | 137
  | 138
  | 139
  | 140
  | 141
  | 142
  | 143
  | 144
  | 145
  | 146
  | 147
  | 148
  | 149
  | 150
  | 151
  | 152
  | 153
  | 154
  | 155
  | 156
  | 157
  | 158
  | 159
  | 160
  | 161
  | 162
  | 163
  | 164
  | 165
  | 166
  | 167
  | 168
  | 169
  | 170
  | 171
  | 172
  | 173
  | 174
  | 175
  | 176
  | 177
  | 178
  | 179
  | 180
  | 181
  | 182
  | 183
  | 184
  | 185
  | 186
  | 187
  | 188
  | 189
  | 190
  | 191
  | 192
  | 193
  | 194
  | 195
  | 196
  | 197
  | 198
  | 199
  | 200
  | 201
  | 202
  | 203
  | 204
  | 205
  | 206
  | 207
  | 208
  | 209
  | 210
  | 211
  | 212
  | 213
  | 214
  | 215
  | 216
  | 217
  | 218
  | 219
  | 220
  | 221
  | 222
  | 223
  | 224
  | 225
  | 226
  | 227
  | 228
  | 229
  | 230
  | 231
  | 232
  | 233
  | 234
  | 235
  | 236
  | 237
  | 238
  | 239
  | 240
  | 241
  | 242
  | 243
  | 244
  | 245
  | 246
  | 247
  | 248
  | 249
  | 250
  | 251
  | 252
  | 253
  | 254
  | 255

export const Uint8Size = 8

/**
 * Generates a series of bitmasks in little-endian order.
 *
 * @example
 * 
 * makeMasks(0) // []
 * makeMasks(1) // [0b1]
 * makeMasks(2) // [0b10, 0b01]
 * makeMasks(3) // [0b100, 0b010, 0b001]
 */
function makeMasks<T extends number>(count: number): T[] {
  const results: T[] = []

  for (let i = count - 1; i >= 0; i--) {
    results.push((1 << i) as T)
  }

  return results
}

export class BitCursor {
  private offset = 0

  public next(): void {
    this.offset += 1
  }

  public prev(): void {
    this.offset -= 1
  }

  public get isByteStart(): boolean {
    return this.offset % Uint8Size === 0
  }

  public get bitOffset(): Uint8Index {
    return (this.offset % Uint8Size) as Uint8Index
  }

  public get byteOffset(): number {
    return (this.offset - this.bitOffset) / Uint8Size
  }

  public mask(bit: Uint1 = 1): Uint8 {
    return bit ? BitCursor.uint8Masks[this.bitOffset] : 0
  }

  public static uint8Masks: readonly Uint8[] = makeMasks(Uint8Size)
}

/**
 * Coerces `number` to a `Uint8` by asserting it is in the range of `Uint8`.
 *
 * @throws if `number` is outside the range of `Uint8` or is not an integer
 */
export function toUint8(number: number): Uint8 {
  if (number < 0x00 || number > 0xff || (number | 0) !== number) {
    throw new TypeError(`cannot convert number to Uint8: ${number}`)
  }
  return number as Uint8
}

/**
 * Gets the size in bits of a number.
 * 
 * @param number a non-negative integer
 * @returns the minimum number of bits required to represent `number`
 * @throws if `number` is negative or is not an integer
 */
export function sizeof(number: number): number {
  if (number < 0 || (number | 0) !== number) {
    throw new TypeError(`cannot get size of negative or non-integer: ${number}`)
  }

  let maxBits = 1

  // eslint-disable-next-line no-cond-assign
  while ((number >>= 1)) {
    maxBits += 1
  }

  return maxBits
}

/**
 * Encoding to be used for encoding and decoding text.
 *
 * @see `BitWriter#writeString`
 * @see `BitReader#readString`
 */
export interface Encoding {
  readonly bitsPerElement: number
  encode(string: string): Uint8Array
  decode(data: Uint8Array): string
}

/**
 * Default encoding to use for `BitReader` and `BitWriter`.
 */
export const UTF8Encoding: Encoding = {
  bitsPerElement: Uint8Size,

  /**
   * Encodes a string as UTF-8 code points.
   */
  encode(string: string): Uint8Array {
    return new TextEncoder('utf-8').encode(string)
  },

  /**
   * Decodes a string from UTF-8 code points.
   */
  decode(data: Uint8Array): string {
    return new TextDecoder('utf-8').decode(data)
  },
}

/**
 * Encoding based on a string of representable characters. Each character is
 * represented by its index within the string, and any other characters are
 * unrepresentable.
 *
 * @example
 *
 * const encoding = new CustomEncoding('0123456789')
 * encoding.encode('23')  // Uint8Array [2, 3]
 */
export class CustomEncoding implements Encoding {
  private readonly chars: string

  /**
   * The maximum character code representable by this class.
   */
  public static MAX_CODE = 1 << (Uint8Array.BYTES_PER_ELEMENT * Uint8Size) - 1
  public readonly bitsPerElement: number

  /**
   * @param chars a string of representable characters without duplicates
   */
  public constructor(chars: string) {
    this.validateChars(chars)
    this.chars = chars
    this.bitsPerElement = sizeof(chars.length - 1)
  }

  private validateChars(chars: string): void {
    if (chars.length > CustomEncoding.MAX_CODE + 1) {
      throw new Error(
        `character set too large, has ${chars.length} but only ${CustomEncoding.MAX_CODE + 1} are allowed`
      )
    }

    for (let i = 0; i < chars.length - 1; i++) {
      const duplicateIndex = chars.indexOf(chars.charAt(i), i + 1)
      if (duplicateIndex > 0) {
        throw new Error(`duplicate character found in character set:\n- set: ${JSON.stringify(chars)}\n- duplicates: ${i} & ${duplicateIndex}`)
      }
    }
  }

  /**
   * Encodes `string` as a series of character codes based on the character set
   * provided to the constructor.
   *
   * @param string a string containing only representable characters
   */
  public encode(string: string): Uint8Array {
    const codes = new Uint8Array(string.length)

    for (let i = 0; i < string.length; i++) {
      const char = string.charAt(i)
      const code = this.chars.indexOf(char)

      if (code < 0) {
        throw new Error(`cannot encode unrepresentable character: ${JSON.stringify(char)} (allowed: ${JSON.stringify(this.chars)})`)
      }

      codes.set([code], i)
    }

    return codes
  }

  /**
   * Builds a string by decoding a character from each character code based on
   * the character set provided to the constructor.
   *
   * @param data a series of character codes representing characters in this encoding
   */
  public decode(data: Uint8Array): string {
    return Array.from(data)
      .map((code, i) => {
        if (code >= this.chars.length) {
          throw new Error(`character code out of bounds at index ${i}: ${code}`)
        }

        return this.chars[code]
      })
      .join('')
  }
}

/**
 * Writes structured data into a `Uint8Array`. Data is written in little-endian
 * order.
 */
export class BitWriter {
  private data = new Uint8Array()
  private cursor = new BitCursor()
  private nextByte: Uint8 = 0b00000000

  /**
   * Writes a single bit.
   */
  public writeUint1(uint1: Uint1): this {
    const mask = this.cursor.mask(uint1)
    this.nextByte |= mask
    this.cursor.next()

    if (this.cursor.isByteStart) {
      this.data = Uint8Array.of(...Array.from(this.data), this.nextByte)
      this.nextByte = 0b00000000
    }

    return this
  }

  /**
   * Writes `1` if given `true`, `0` otherwise.
   */
  public writeBoolean(boolean: boolean): this {
    return this.writeUint1(boolean ? 1 : 0)
  }

  /**
   * Writes data from a `Uint8` by writing 8 bits.
   */
  public writeUint8(uint8: Uint8): this {
    return this.writeUint(uint8, { size: Uint8Size })
  }

  /**
   * Writes an unsigned integer as a series of bits. There are two ways to
   * control the number of bytes `number` takes: provide a `max` value or
   * provide a `size`. If `max` is provided, then however many bytes would be
   * required to write `max` will be used to write `number`. If `size` is
   * provided, then that is how many bytes will be used.
   *
   * @example
   *
   * bits.writeUint(23, { max: 30 })  // writes `10111`
   * bits.writeUint(99, { size: 8 })  // writes `01100011`
   */
  public writeUint(
    number: number,
    { max, size }: { max?: number; size?: number } = {}
  ): this {
    if (typeof max !== 'undefined' && typeof size !== 'undefined') {
      throw new Error("cannot specify both 'max' and 'size' options")
    }

    if (typeof max !== 'undefined') {
      if (number > max) {
        throw new Error(`overflow: ${number} must be less than ${max}`)
      }

      size = sizeof(max)
    }

    if (typeof size === 'undefined') {
      throw new Error('size cannot be undefined')
    }

    if (number >= 1 << size) {
      throw new Error(`overflow: ${number} cannot fit in ${size} bits`)
    }

    for (const mask of makeMasks(size)) {
      this.writeUint1((number & mask) === 0 ? 0 : 1)
    }

    return this
  }

  /**
   * Writes string data with an encoding and maximum length. By default, the
   * encoding used will be UTF-8. If your string has a restricted character set,
   * you can use your own `CustomEncoding` to read and write the string more
   * compactly than you otherwise would be able to.
   * 
   * It is important to remember that the `encoding` and `maxLength` options
   * must be the same for `readString` and `writeString` calls, otherwise
   * reading the string will very likely fail or be corrupt.
   *
   * @example
   *
   * bits.writeString('hi') // uses utf-8, writes '0110100001101001'
   *
   * const encoding = new CustomEncoding('hi')
   * bits.writeString('hi') // uses custom encoding, writes '01'
   */
  public writeString(
    string: string,
    {
      encoding = UTF8Encoding,
      maxLength = 255,
    }: { encoding?: Encoding; maxLength?: number } = {}
  ): this {
    const codes = Array.from(encoding.encode(string))

    if (codes.length > maxLength) {
      throw new Error(
        `overflow: cannot write a string longer than max length: ${string.length} > ${maxLength}`
      )
    }

    // write length
    this.writeUint(codes.length, { max: maxLength })

    // write content
    for (const code of codes) {
      this.writeUint(code, { size: encoding.bitsPerElement })
    }

    return this
  }

  /**
   * Converts the data written to this `BitWriter` to a `Uint8Array`.
   */
  public toUint8Array(): Uint8Array {
    const pendingByte = this.getPendingByte()
    return Uint8Array.of(
      ...Array.from(this.data),
      ...(typeof pendingByte !== 'undefined' ? [pendingByte] : [])
    )
  }

  /**
   * If there's a byte that is not yet full, get it.
   */
  private getPendingByte(): Uint8 | undefined {
    if (!this.cursor.isByteStart) {
      return this.nextByte
    }
  }
}

/**
 * Reads structured data from a `Uint8Array`. Data is read in little-endian
 * order.
 */
export class BitReader {
  private data: Uint8Array
  private cursor = new BitCursor()

  /**
   * @param data a buffer to read data from
   */
  public constructor(data: Uint8Array) {
    this.data = data
  }

  /**
   * Reads a Uint1 and moves the internal cursor forward one bit.
   */
  public readUint1(): Uint1 {
    const byte = this.getCurrentByte()
    const mask = this.cursor.mask()

    this.cursor.next()

    return (byte & mask) === 0 ? 0 : 1
  }

  /**
   * Reads a number by reading 8 bits.
   */
  public readUint8(): Uint8 {
    return this.readUint({ size: Uint8Size }) as Uint8
  }

  /**
   * Reads a boolean by reading a bit and returning whether the bit was set.
   */
  public readBoolean(): boolean {
    return this.readUint1() === 1
  }

  /**
   * Reads a unsigned integer as a series of bits. There are two ways to control
   * the number of bytes `number` takes: provide a `max` value or provide a
   * `size`. If `max` is provided, then however many bytes would be required to
   * write `max` will be used to write `number`. If `size` is provided, then
   * that is how many bytes will be used.
   *
   * @example
   *
   * // contains 16 bits
   * const bits = new BitReader(Uint8Array.of(0xff, 0x0f))
   *
   * bits.readUint({ max: 0x0f })  // reads first 4 bits: 0x0f
   * bits.readUint({ size: 8 })    // reads next 8 bits:  0xf0
   * bits.readUint({ size: 4 })    // reads last 4 bits:  0x0f
   */
  public readUint({max, size }: { max?: number, size?: number } = {}): number {
    if (typeof max !== 'undefined' && typeof size !== 'undefined') {
      throw new Error("cannot specify both 'max' and 'size' options")
    }

    if (typeof max !== 'undefined') {
      size = sizeof(max)
    }

    if (typeof size === 'undefined') {
      throw new Error('size cannot be undefined')
    }

    let result = 0

    for (const mask of makeMasks(size)) {
      const bit = this.readUint1()
      if (bit) {
        result |= mask
      }
    }

    return result
  }

  /**
   * Reads a length-prefixed string with an encoding and maximum length. By
   * default, the encoding used will be UTF-8. If your string has a restricted
   * character set, you can use your own `CustomEncoding` to read and write the
   * string more compactly than you otherwise would be able to.
   *
   * It is important to remember that the `encoding` and `maxLength` options
   * must be the same for `readString` and `writeString` calls, otherwise
   * reading the string will very likely fail or be corrupt.
   *
   * @example
   *
   *                                  // length  'h'  'i'
   *                                  //      ↓   ↓    ↓
   * const bits = new BitReader(Uint8Array.of(2, 104, 105))
   * bits.readString() // "hi"
   *
   *                                  // length  'h''i'
   *                                  //           ↓↓
   * const bits = new BitReader(Uint8Array.of(2, 0b01000000))
   * const encoding = new CustomEncoding('hi')    // ↑↑↑↑↑↑
   * bits.readString() // "hi"                       padding
   */
  public readString({
    encoding = UTF8Encoding,
    maxLength = 255,
  }: { encoding?: Encoding, maxLength?: number } = {}): string {
    const length = this.readUint({ max: maxLength })
    const codes = new Uint8Array(length)

    for (let i = 0; i < length; i++) {
      codes.set([this.readUint({ size: encoding.bitsPerElement })], i)
    }

    return encoding.decode(codes)
  }

  /**
   * Determines whether there is any more data to read. If the result is
   * `false`, then any call to read data will throw an exception.
   */
  public canRead(): boolean {
    return this.cursor.byteOffset < this.data.length
  }

  private getCurrentByte(): Uint8 {
    if (this.cursor.byteOffset >= this.data.length) {
      throw new Error(
        `end of buffer reached: byteOffset=${this.cursor.byteOffset} bitOffset=${this.cursor.bitOffset} data.length=${this.data.length}`
      )
    }

    return toUint8(this.data[this.cursor.byteOffset])
  }
}
