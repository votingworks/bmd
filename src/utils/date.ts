import moment from 'moment'

export const AMERICA_TIMEZONES = [
  'Pacific/Honolulu',
  'America/Anchorage',
  'America/Los_Angeles',
  'America/Phoenix',
  'America/Denver',
  'America/Chicago',
  'America/New_York',
]

function* getShortMonthNames(): Generator<string> {
  const monthShortNameFormatter = new Intl.DateTimeFormat(undefined, {
    month: 'short',
  })
  const year = new Date().getFullYear()
  for (
    let month = 0;
    new Date(year, month, 1).getFullYear() === year;
    month++
  ) {
    yield monthShortNameFormatter.format(new Date(year, month, 1))
  }
}
export const MONTHS_SHORT = [...getShortMonthNames()]

export const formatTimeZoneName = (date: Date, timeZone?: string) =>
  new Intl.DateTimeFormat(undefined, {
    timeZoneName: 'long',
    timeZone,
  })
    .formatToParts(date)
    .find((part) => part.type === 'timeZoneName')?.value

export const formatFullDateTimeZone = (date: Date, timeZone?: string) =>
  new Intl.DateTimeFormat(undefined, {
    weekday: 'short',
    month: 'long',
    day: 'numeric',
    year: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    timeZoneName: 'short',
    timeZone,
  }).format(date)

export const formatShortWeekdayLongDate = (date: Date) =>
  new Intl.DateTimeFormat(undefined, {
    weekday: 'short',
    month: 'long',
    day: 'numeric',
    year: 'numeric',
  }).format(date)

export const formatShortTime = (date: Date) =>
  new Intl.DateTimeFormat(undefined, {
    hour: 'numeric',
    minute: 'numeric',
  }).format(date)

export const formatLongDate = (date: Date) =>
  new Intl.DateTimeFormat(undefined, {
    month: 'long',
    day: 'numeric',
    year: 'numeric',
  }).format(date)

// TODO: replace `dateLong` with `formatLongDate`
export const dateLong = (dateString: string) =>
  moment(new Date(dateString)).format('LL')

/**
 * Get days in given month and year.
 *
 * @param year
 * @param month
 *
 * @return Date[]
 *
 */
export function getDaysInMonth(year: number, month: number) {
  const date = new Date(year, month, 1)
  const days = []
  while (date.getMonth() === month) {
    days.push(new Date(date))
    date.setDate(date.getDate() + 1)
  }
  return days
}
