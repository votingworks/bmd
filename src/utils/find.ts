import { Parties } from '@votingworks/ballot-encoder'

export const findPartyById = (parties: Parties, id: string) => {
  return parties.find((p) => p.id === id)
}

export default { findPartyById }
