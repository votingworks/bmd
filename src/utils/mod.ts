const mod = (x: number, n: number) => ((x % n) + n) % n
export default mod
