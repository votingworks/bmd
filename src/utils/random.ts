/**
 * Random Base64 string.
 * Get random values to fill an array of prescribed length,
 * b64 encode it, and strip the '=' at the end.
 *
 * @param numBytes Number of bytes
 *
 * @return string
 *
 */

export function randomBase64(numBytes: number) {
  return window
    .btoa(
      String.fromCharCode(
        ...window.crypto.getRandomValues(new Uint8ClampedArray(numBytes))
      )
    )
    .replace(/=+$/, '')
}

export default { randomBase64 }
