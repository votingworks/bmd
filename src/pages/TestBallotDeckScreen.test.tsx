import {
  //electionSample,
  Election,
} from '@votingworks/ballot-encoder'
// TODO: Tally: Use electionSample from ballot-encoder once published.

import React from 'react'
import { fireEvent } from '@testing-library/react'
import electionSample from '../data/electionSample.json'
import { mockOf, render } from '../../test/testUtils'
import { randomBase64 } from '../utils/random'
import TestBallotDeckScreen from './TestBallotDeckScreen'

import fakeKiosk from '../../test/helpers/fakeKiosk'

// mock the random value so the snapshots match
jest.mock('../utils/random')
const randomBase64Mock = mockOf(randomBase64)
randomBase64Mock.mockReturnValue('CHhgYxfN5GeqnK8KaVOt1w')

it('renders test decks appropriately', () => {
  const { getAllByText, getByText } = render(
    <TestBallotDeckScreen
      appName="VxPrint"
      appPrecinctId="23"
      election={electionSample as Election}
      hideTestDeck={jest.fn()}
      isLiveMode={false}
    />
  )

  fireEvent.click(getByText('All Precincts'))

  expect(getAllByText('Unofficial TEST Ballot')).toHaveLength(63)
  expect(getAllByText('For either', { exact: false })).toHaveLength(31)
  expect(getAllByText('FOR Measure 420A', { exact: false })).toHaveLength(31)
  expect(getAllByText('County Commissioners')).toHaveLength(52)

  const oldWindowPrint = window.print
  delete window.print
  window.print = jest.fn()

  fireEvent.click(getByText('Print 63 ballots'))

  expect(window.print).toHaveBeenCalled()

  window.print = oldWindowPrint

  window.kiosk = fakeKiosk()

  fireEvent.click(getByText('Print 63 ballots'))

  expect(window.kiosk!.print).toHaveBeenCalledWith()
})
