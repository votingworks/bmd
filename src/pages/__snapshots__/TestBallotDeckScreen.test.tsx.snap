// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders test decks appropriately 1`] = `
.c5 {
  border: none;
  border-radius: 0.25rem;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.35rem 0.5rem;
  line-height: 1.25;
  color: black;
  touch-action: manipulation;
}

.c4 {
  border: none;
  border-radius: 0.25rem;
  box-shadow: 0 0 0 0 rgba(71,167,75,1);
  box-sizing: border-box;
  background: rgb(71,167,75);
  cursor: pointer;
  padding: 1rem 1.75rem;
  line-height: 1.25;
  color: #FFFFFF;
  font-size: 1.25rem;
  touch-action: manipulation;
}

.c14 {
  max-width: 250px;
}

.c15 {
  max-width: 250px;
}

.c3 {
  max-width: 66ch;
  line-height: 1.2;
}

.c3 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c3 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c3 h3 {
  font-size: 1.17rem;
}

.c3 h4 {
  font-size: 1rem;
}

.c3 h5 {
  font-size: 0.9rem;
}

.c3 h3,
.c3 h4,
.c3 h5,
.c3 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p {
  margin-top: -0.75rem;
}

.c3 h3 + p,
.c3 h4 + p,
.c3 h5 + p {
  margin-top: -1rem;
}

.c3 > :first-child {
  margin-top: 0;
}

.c3 > :last-child {
  margin-bottom: 0;
}

.c3 dl {
  margin: 1rem 0;
}

.c16 {
  max-width: 66ch;
  line-height: 1.2;
}

.c16 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c16 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c16 h3 {
  font-size: 1.17rem;
}

.c16 h4 {
  font-size: 1rem;
}

.c16 h5 {
  font-size: 0.9rem;
}

.c16 h3,
.c16 h4,
.c16 h5,
.c16 p {
  margin-top: 0;
  margin-bottom: 0;
}

.c16 h1 + h2 {
  margin-top: -0.75rem;
}

.c16 h1 + p,
.c16 h2 + p {
  margin-top: -0.75rem;
}

.c16 h3 + p,
.c16 h4 + p,
.c16 h5 + p {
  margin-top: 0;
}

.c16 > :first-child {
  margin-top: 0;
}

.c16 > :last-child {
  margin-bottom: 0;
}

.c16 dl {
  margin: 1rem 0;
}

.c8 {
  text-align: center;
}

.c9 {
  white-space: nowrap;
  font-weight: 300;
}

.c17 {
  font-size: 0.8rem;
}

.c19 {
  font-size: 0.8rem;
  font-weight: 300;
}

.c18 {
  white-space: nowrap;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 1.5rem 1rem 0;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: auto;
}

.c13 div:first-child {
  margin-right: 1rem;
  min-width: 5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
  padding: 1rem 0.5rem 2rem;
}

.c2 {
  margin: 0 0;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #333333;
  color: #ffffff;
}

.c7 {
  margin: 2rem 1rem 1rem;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 auto;
  max-width: 20rem;
  padding: 1rem;
}

.c10 p > button {
  width: 100%;
}

.c11 {
  margin: 1rem 1rem 1.5rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  height: 100%;
}

.c0 > nav {
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

.c0 > main {
  -webkit-flex: 3;
  -ms-flex: 3;
  flex: 3;
}

@media (min-width:480px) {
  .c3 {
    line-height: 1.3;
  }
}

@media (min-width:480px) {
  .c16 {
    line-height: 1.3;
  }
}

@media (min-width:480px) {
  .c1 {
    padding: 1rem;
  }
}

@media print {
  .c0 {
    display: none;
  }
}

<div
  class="c0"
>
  <main
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3 no-print"
      >
        <h1>
          Test Ballot Decks
        </h1>
        <p>
          Deck containing
           
          <strong>
            63 ballots
          </strong>
           
          for 
          
          .
        </p>
        <p>
          <button
            class="c4"
            type="button"
          >
            Print 
            63
             ballots
          </button>
        </p>
        <p>
          <button
            class="c5"
            type="button"
          >
            Back to Precincts List
          </button>
        </p>
      </div>
    </div>
  </main>
  <nav
    class="c6"
  >
    <div
      class="c7"
    >
      <div
        class="c3"
      >
        <h1
          class="c8"
        >
          VxPrint
          <span
            class="c9"
          >
             / 
            Election Admin Actions
          </span>
        </h1>
      </div>
    </div>
    <div
      class="c10"
    >
      <button
        class="c5"
        type="button"
      >
        Back to Admin Dashboard
      </button>
    </div>
    <div
      class="c11"
    >
      <div
        aria-hidden="true"
        class="c12"
        data-testid="election-info"
      >
        <div
          class="c13"
        >
          <div
            aria-hidden="true"
            class="c14"
          >
            <img
              alt="state seal"
              class="c15"
              src="/seals/Sample-Seal.svg"
            />
          </div>
          <div
            class="c16"
          >
            <h5
              aria-label=" General Election."
            >
               General Election
            </h5>
            <p
              class="c17"
            >
              November 3, 2020
              <br />
              <span
                class="c18"
              >
                Franklin County
                ,
              </span>
               
              <span
                class="c18"
              >
                State of Hamilton
              </span>
            </p>
            <p
              class="c19"
            >
              <span
                class="c18"
              >
                Center Springfield
              </span>
               
            </p>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
